Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,726 @@
+
+        if (!visitor(entry)) {
+            return RecursionDecision::Break;
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error())
+        return traversal_result;
+
+    return KSuccess;
+}
+
+RefPtr<Inode> ISO9660Inode::lookup(StringView name)
+{
+    auto& file_system = static_cast<ISO9660FS const&>(fs());
+    RefPtr<Inode> inode;
+    Array<u8, max_file_identifier_length> file_identifier_buffer;
+
+    auto traversal_result = file_system.visit_directory_record(&m_record, [&](ISO::DirectoryRecordHeader const* record) {
+        StringView filename = get_normalized_filename(record, file_identifier_buffer);
+
+        if (filename == name) {
+            auto maybe_inode = ISO9660Inode::create_from_directory_record(const_cast<ISO9660FS&>(file_system), record, filename);
+            if (maybe_inode.is_error()) {
+                // FIXME: The Inode API does not handle allocation failures very
+                //        well... we can't return a KResultOr from here. It
+                //        would be nice if we could return a KResult(Or) from
+                //        any place where allocation may happen.
+                dbgln("Could not allocate inode for lookup!");
+            } else {
+                inode = maybe_inode.release_value();
+            }
+            return RecursionDecision::Break;
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error()) {
+        return {};
+    }
+
+    return inode;
+}
+
+void ISO9660Inode::flush_metadata()
+{
+}
+
+KResultOr<size_t> ISO9660Inode::write_bytes(off_t, size_t, const UserOrKernelBuffer&, FileDescription*)
+{
+    return EROFS;
+}
+
+KResultOr<NonnullRefPtr<Inode>> ISO9660Inode::create_child(StringView, mode_t, dev_t, uid_t, gid_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::add_child(Inode&, const StringView&, mode_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::remove_child(const StringView&)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::chmod(mode_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::chown(uid_t, gid_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::truncate(u64)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_atime(time_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_ctime(time_t)
+{
+    return EROFS;
+}
+
+KResult ISO9660Inode::set_mtime(time_t)
+{
+    return EROFS;
+}
+
+void ISO9660Inode::one_ref_left()
+{
+}
+
+ISO9660Inode::ISO9660Inode(ISO9660FS& fs, ISO::DirectoryRecordHeader const* record, StringView const& name)
+    : Inode(fs, get_inode_index(record, name))
+    , m_record(*record)
+{
+    dbgln_if(ISO9660_VERY_DEBUG, "Creating inode #{}", index());
+    create_metadata();
+}
+
+ISO9660Inode::~ISO9660Inode()
+{
+}
+
+KResultOr<NonnullRefPtr<ISO9660Inode>> ISO9660Inode::create_from_directory_record(ISO9660FS& fs, ISO::DirectoryRecordHeader const* record, StringView const& name)
```

Comment:
This function does not support `record` being null, so it should take `record` as a reference.